### PR TITLE
add error-checked version of walkDir

### PR DESCRIPTION
**Pull request by a-mr**
Alternative solution w.r.t. #13011, #12960.
New iterator `walkDirErr` with error-checking, use like this:
```nim
for kind, path, error in walkDirErr("dirA"):
  if error == OSErrorCode(0):
    echo path & " of kind " & $kind
  else:
    echo path & " GET INFO FAILED: " & osErrorMsg(error)
```
Other iterators remain unchecked.
Tested manually on Linux & Windows.

### CODE REVIEW

**Comment by juancarlospaco** [comment]
Being Nim a typed lang with overload and all, I dont think we need to add `functionErrorname`.
`if err != nil: ...` feels too Go lang.

**Comment by Araq** [comment]
I like this idea much better than https://github.com/nim-lang/Nim/pull/13011

**Comment by Araq** [comment]
Needs a ``.since`` annotation and a changelog entry.

**Comment by juancarlospaco** [comment]
What about `tryWalkDir` name to be consistent with the rest of the std lib ?.

**Comment by timotheecour** [comment]
will `walkDirRec` get similar treatment?

**Comment by Araq** [comment]
Indeed, please rename it to ``tryWalkDir``.

**Comment by demotomohiro** [comment]
When `findFirstFile` or `opendir` failed (because `dir` is invalid path or no permission), `walkDirErr` yield `dir`.
But when there is no error, it doesn't yield `dir`.
How about to add `pcInputPath` to `PathComponent` enum and yield it in that case?
I can check if opening input `dir` is failed by comparing `path` and `dir`, but just checking `error` and `kind` is faster.

Can you write signle yield per when branch code?
https://nim-lang.org/docs/manual.html#iterators-and-the-for-statement-first-class-iterators
> the body of a for loop over an inline iterator is inlined into each yield statement appearing in the iterator code, so ideally the code should be refactored to contain a single yield when possible to avoid code bloat.

File: lib/pure/os.nim [hunk 1]
```diff
@@ -2025,14 +2010,16 @@ iterator walkDir*(dir: string; relative=false): tuple[kind: PathComponent, path:
               k = succ(k)
             let xx = if relative: extractFilename(getFilename(f))
                      else: dir / extractFilename(getFilename(f))
-            yield (k, xx)
+            yield (k, xx, OSErrorCode(0))
           if findNextFile(h, f) == 0'i32:
             let errCode = getLastError()
             if errCode == ERROR_NO_MORE_FILES: break
             else: raiseOSError(errCode.OSErrorCode)
```
Comment (demotomohiro) [inline lib/pure/os.nim]: There is still a case it raise error when `findNextFile` failed.

File: lib/pure/os.nim [hunk 2]
```diff
@@ -2025,14 +2010,16 @@ iterator walkDir*(dir: string; relative=false): tuple[kind: PathComponent, path:
...
     else:
       var d = opendir(dir)
-      if d != nil:
+      if d == nil:
+        yield(pcFile, dir, osLastError())
+      else:
         defer: discard closedir(d)
         while true:
           var x = readdir(d)
```
Comment (demotomohiro) [inline lib/pure/os.nim]: According to man page of `readdir`,
https://linux.die.net/man/3/readdir
> If an error occurs, NULL is returned and errno is set appropriately.

**Comment by Araq** [comment]
> How about to add pcInputPath to PathComponent enum and yield it in that case?

That would break code, code can contain exhaustive case statements, in fact, I generally encourage people to do that.

> will walkDirRec get similar treatment?

No, we only need one non-recursive iterator as the building block for custom walks that can live outside the stdlib.

**Comment by a-mr** [comment]
For discussion: an updated version, slightly more generalized + some fixes:
```nim
for step in tryWalkDir(dir):
  case step.kind
  of wiOpenDir:
    if step.openStatus != odOpenOk:
      echo "directory " & dir & " could not be opened " &
           $step.openStatus & " / " & osErrorMsg(step.code)
  of wiEntryOk: echo "found " & step.path & " of kind " & $step.entryType
  of wiEntryBad: echo "bad symlink " & step.path & " with error " &
                      osErrorMsg(step.code)
  of wiInterrupted:
    echo "traversal interrupted with error: " & osErrorMsg(step.code)
```

**Review by demotomohiro — APPROVED** [review]

File: lib/pure/os.nim (hunk 2 above)
Comment (a-mr) [reply lib/pure/os.nim]: addressed in new version

File: lib/pure/os.nim [hunk 3]
```diff
@@ -2088,6 +2088,216 @@ iterator walkDir*(dir: string; relative=false): tuple[kind: PathComponent, path:
...
+  ##     case step.openStatus:
+  ##     of odOpenOk:       echo dir & " - directory is opened OK"
+  ##     of odNotDir:       echo dir & " is not a directory" & err
+  ##     of odAccessDenied: echo dir & " - Access Denied" & err
+  ##     of odNotFound:     echo dir & " - no such path" & err
+  ##     of odUnknownError: echo dir & "unknown error - that's bad!" & err
+  ##   of wsEntryOk: echo step.path & " is entry of kind " & $step.entryType
+  ##   of wsEntryBad: echo step.path & " is broken symlink " & err
+  ##   of wsInterrupted:
+  ##     echo dir & " traversal interrupted, really bad! " & err
```
Comment (dom96) [inline lib/pure/os.nim]: Indentation is wrong here

File: lib/pure/os.nim [hunk 4]
```diff
@@ -2088,6 +2088,216 @@ iterator walkDir*(dir: string; relative=false): tuple[kind: PathComponent, path:
...
+  ## thorough error checking. It never raises an exception.
+  ## Yields *steps* of walking through `dir`, each step contains
+  ## its path ``path``, OS error code ``code`` and additional tag ``kind``.
+  ## At first step it yields info ``openStatus`` of opening
+  ## status `dir` as a directory, tag ``wsOpenDir``.
+  ## If it's OK, the subsequent steps will yield the directory entries
+  ## with component type as info ``entryType``, tag ``wsEntryOk``.
+  ## Dangling symlinks on posix are reported as ``wsEntryBad``.
+  ## If (rarely) walking terminates with an error (such as an I/O error),
+  ## tag is `wsInterrupted`.
```
Comment (dom96) [inline lib/pure/os.nim]: Please reword this to not be a wall of text.

Split into paragraphs:

1. Short description of what this is.
2. Longer description of how it works.
3. Edge cases/other details.

File: lib/pure/os.nim [hunk 5]
```diff
@@ -2088,6 +2088,216 @@ iterator walkDir*(dir: string; relative=false): tuple[kind: PathComponent, path:
...
+  var lastIter = false
+  when defined(windows):
+    var yieldAllowed: bool
+    var nStep = 1
+    var h: Handle = -1
+    defer:
+      if h != -1: findClose(h)
+    var f: WIN32_FIND_DATA
+    template resolveFile() =
+        var k = pcFile
```
Comment (dom96) [inline lib/pure/os.nim]: Fix the inconsistent indentation.

File: lib/pure/os.nim [hunk 6]
```diff
@@ -2088,6 +2088,216 @@ iterator walkDir*(dir: string; relative=false): tuple[kind: PathComponent, path:
               k = getSymlinkFileKind(path)
             yield (k, y)
 
+type OpenDirStatus* = enum
```
Comment (dom96) [inline lib/pure/os.nim]: convention is to put `type` on its own line

**Review by dom96 — CHANGES_REQUESTED** [review]
I left a few comments, but **I don't think this should be accepted into the stdlib** for various reasons:

* I don't see this seeing enough use to be included in the stdlib
* The API for this should be modelled via something more generic, like a `Either[T, E]` type or something instead of this custom WalkStep object.
* The code being introduced in this PR is convoluted and hard to read, which will make it hard to maintain

File: lib/pure/os.nim (hunk 6 above)
Comment (timotheecour) [reply lib/pure/os.nim]: when the block contains a single type, it's quite common to write
`type Foo = ...` instead of 
```nim
type
  Foo = ...
```
and is in fact easier to search.
the main advantage of the latter form is for mutually recursive types

**Comment by a-mr** [comment]
The solution of using closure iterators would be much simpler and more elegant (return `tuple[OpenDirStatus, iterator]`), but I'm not yet sure how to manage resources for them.
It seems that directory handles, captured by closure iterators are never freed.
A naive attempt to define a hook for the handles does not even pass compilation:
```nim
proc `=destroy`(m: var ptr DIR) =
  echo "Destruct"

Error: signature for '=destroy' must be proc[T: object](x: var T)
```
I can try using wrapper object for the handle.
There is a recent discussion on topic but it's without conclusion: [Will --gc:arc address finalizers for closure iterator variables?](https://forum.nim-lang.org/t/5845)

File: lib/pure/os.nim (hunk 6 above)
Comment (dom96) [reply lib/pure/os.nim]: It's common, but not in Nim's codebase.

**Comment by Araq** [comment]
> I don't see this seeing enough use to be included in the stdlib

I've asked for an implementation like this because our current means are insufficient for handling errors.

File: lib/pure/os.nim (hunk 3 above)
Comment (a-mr) [reply lib/pure/os.nim]: fixed, thx

File: lib/pure/os.nim (hunk 5 above)
Comment (a-mr) [reply lib/pure/os.nim]: fixed

File: lib/pure/os.nim (hunk 4 above)
Comment (a-mr) [reply lib/pure/os.nim]: I tried to improve that.

File: lib/pure/os.nim (hunk 6 above)
Comment (a-mr) [reply lib/pure/os.nim]: OK

**Comment by a-mr** [comment]
It looks a bit bulky but there is a bright side: recursive version is straightforward for implementation, because `WalkStep` values can be passed upstream easily. Here is an analogue for walkDirRec from stdlib:

```nim
iterator tryWalkDirRec*(dir: string,
                        yieldFilter = {pcFile}, followFilter = {pcDir},
                        relative = false):
                       WalkStep {.tags: [ReadDirEffect].} =
  var stack = @[""]
  while stack.len > 0:
    let d = stack.pop()
    for s in tryWalkDir(dir / d, relative = true):
      let rel = d / s.path
      case s.kind
      of wsEntryOk:
        if s.entryType in {pcDir, pcLinkToDir} and
           s.entryType in followFilter:
          stack.add rel
        if s.entryType in yieldFilter:
          var step = s
          step.path = if relative: rel else: dir / rel
          yield step
      of wsEntryBad, wsInterrupted, wsOpenDir:
        var step = s
        step.path = if relative: rel else: dir / rel
        yield step
```

**Comment by timotheecour** [comment]
multiple yield statements in inline iterators cause code bloat and must be avoided, as explained here: https://github.com/nim-lang/Nim/pull/10553

see https://github.com/nim-lang/Nim/pull/10557 (which was closed but is relevant) for an example of how to convert multiple yield into a single yield

both current PR and https://github.com/nim-lang/Nim/pull/13163#issuecomment-582160275 suffer from this

One possibility would be to use a 1st class iterator instead of inline iterator, given cost of file system operations, the overhead of 1st class iterators would be negligible

**Comment by a-mr** [comment]
it was rewritten to use single yield. I'm using multiple yield only in examples for the sake of clarity.

File: lib/pure/os.nim [hunk 7]
```diff
@@ -2088,6 +2095,221 @@ iterator walkDir*(dir: string; relative=false): tuple[kind: PathComponent, path:
...
+      errno = 0
+      var x = readdir(d)
+      if x == nil:
+        if errno == 0:
+          break  # normal end, yielding nothing
+        else:
+          skip = false
+          step = sGetError(wsInterrupted, outDir)
+          continue
+      when defined(nimNoArrayToCstringConversion):
```
Comment (Araq) [inline lib/pure/os.nim]: This ``when`` is always true.

File: lib/pure/os.nim [hunk 8]
```diff
@@ -2088,6 +2095,221 @@ iterator walkDir*(dir: string; relative=false): tuple[kind: PathComponent, path:
...
+        else:
+          return sGetError(wsEntryBad, y)
+
+    var d: ptr DIR = opendir(dir)
+    defer:
+      if d != nil:
+        discard closedir(d)
+    let status =
+      if d == nil:
+          if errno == ENOENT: odNotFound
```
Comment (Araq) [inline lib/pure/os.nim]: Indentation.

File: lib/pure/os.nim [hunk 9]
```diff
@@ -2088,6 +2095,221 @@ iterator walkDir*(dir: string; relative=false): tuple[kind: PathComponent, path:
...
+    step = sOpenDir(status, outDir)
+
+    while true:
+      if not skip:
+        yield step
+      if d == nil or step.kind == wsInterrupted:
+        break
+      let errnoSave = errno
+      defer: errno = errnoSave
+      errno = 0
```
Comment (Araq) [inline lib/pure/os.nim]: Suspicious code.

**Comment by Araq** [comment]
This definitely goes in the right direction, but:

1. The control flow is unstructured. Too many ``break`` and ``continue`` statements make it hard to reason about. I personally also hate ``defer``.
2. No less than 4 different enums for iterating over a dir? OpenDirStatus, WalkStepKind, OSErrorCode          , PathComponent?

I understand that iterating over directories properly in 2020 involves some black magic, but this seems overly excessive.

File: lib/pure/os.nim [hunk 10]
```diff
@@ -2088,6 +2095,221 @@ iterator walkDir*(dir: string; relative=false): tuple[kind: PathComponent, path:
...
+    code*: OSErrorCode           ## OS error code for the step
+    case kind*: WalkStepKind     ## discriminator
+    of wsOpenDir:
+      openStatus*: OpenDirStatus ## status of opening the directory
+    of wsEntryOk:
+      entryType*: PathComponent  ## path component
+    of wsEntryBad, wsInterrupted:
+      discard
+
+proc `$`*(s: WalkStep): string =
```
Comment (timotheecour) [inline lib/pure/os.nim]: this results in non-standard formatting, ideally should look like builtin `$` for obj (in particular, double quotes not single quotes for strings);
maybe all that's needed is to add `$` for OSErrorCode and then the builtin `$` would just work

File: lib/pure/os.nim (hunk 7 above)
Comment (a-mr) [reply lib/pure/os.nim]: fixed

File: lib/pure/os.nim (hunk 8 above)
Comment (a-mr) [reply lib/pure/os.nim]: fixed

File: lib/pure/os.nim (hunk 9 above)
Comment (a-mr) [reply lib/pure/os.nim]: fixed, preserving errno is done more correct way, without defer.

File: lib/pure/os.nim (hunk 10 above)
Comment (a-mr) [reply lib/pure/os.nim]: I don't know why it did not work, strange. Now, after I united openDirStatus and WalkStep enums it started to work as a bultin indeed.

**Comment by a-mr** [comment]
I combined OpenDirStatus into WalkStepKind. And I tried to simplify the code further, with limited success though. It's hard to do that with inline iterator using single yield statement.  
It seems I've reached my limit here :-/

**Comment by a-mr** [comment]
@Araq,  
We need at least one custom enum to decode OS errors, using only OSErrorCode (which is not even enum) does not look good enough, it would be more like shifting responsibility to users.

I actually support the idea to introduce Either type,  as @dom96 suggested. Well it will not simplify things for this particular PR, but eventually it will be very beneficial, becoming a pattern for any computations that may fail. It's plain strange that Nim does not offer any official pattern for passing errors. It has exceptions, but they are falling out of fashion, so what to use instead?

**Comment by Araq** [comment]
> It has exceptions, but they are falling out of fashion, so what to use instead?

Exceptions in general, because they will be fashionalbe once again. When people realize that the alternatives suck for program composition and code evolvability. Doesn't help this case though as for this case the control flow exceptions introduce is wrong.

**Comment by a-mr** [comment]
@Araq, @timotheecour, @dom96,
Thanks for feedback.
I finished my corrections according to your comments.

**Comment by timotheecour** [comment]
> I combined OpenDirStatus into WalkStepKind. And I tried to simplify the code further, with limited success though. It's hard to do that with inline iterator using single yield statement.
It seems I've reached my limit here :-/

filewalks (https://github.com/nim-lang/fusion/pull/32) does this; IMO filewalks belongs squarely in stdlib.

**Comment by a-mr** [comment]
IMHO [nim-results](https://github.com/arnetheduck/nim-result) should be a part of stdlib for the tasks like this PR.
